File: frontend/src/stores/blueprint.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { agentFetch, agentJson, agentPost } from '@/api'

// ═══════════════════════════════════════════════
// Blueprint Store — 蓝图工作流管理
// ═══════════════════════════════════════════════

export interface SkillItem {
  name: string
  description: string
}

// Backend-compatible node position
export interface BlueprintNodePosition {
  x: number
  y: number
}

// Backend-compatible node data (maps to BlueprintNode.config + label etc.)
export interface BlueprintNodeData {
  label?: string
  prompt?: string
  expression?: string
  max_iterations?: number
  template?: string
  instructions?: string
  port_count?: number
  max_handoffs?: number
  slot?: number
  execution_mode?: string
  parallel_lane_count?: number
  approval?: boolean
  model_id?: string
  skill_ids?: string[]
  tools?: string[]
  runtime_id?: string
  agent_name?: string
  permission_profile?: string
  working_directory?: string
  timeout_ms?: number
  output_schema?: unknown
  disabled?: boolean
  [key: string]: unknown
}

export interface BlueprintNode {
  id: string
  type: 'agent' | 'condition' | 'loop' | 'summary' | 'note' | 'manager' | 'manager_slot' | 'group'
  label: string
  position: BlueprintNodePosition
  data: BlueprintNodeData
  parent_id?: string
  disabled?: boolean
  config?: Record<string, unknown>
}

export interface BlueprintEdge {
  id: string
  source: string
  target: string
  sourceHandle?: string | null
  targetHandle?: string | null
  condition?: string
  label?: string
}

export interface Blueprint {
  id: string
  name: string
  description: string
  nodes: BlueprintNode[]
  edges: BlueprintEdge[]
  created_at: string
  updated_at: string
  version?: number
}

export interface BlueprintRunNodeData {
  status?: string
  output?: string
  error?: string
}

export interface BlueprintRun {
  id: string
  blueprint_id: string
  blueprint_name?: string
  status: 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled' | 'waiting_approval'
  started_at: string
  finished_at?: string
  node_statuses?: Record<string, string>
  error?: string
  nodes?: Record<string, BlueprintRunNodeData>
  node_runs?: Array<{
    node_id: string
    status: string
    output?: string
    error?: string
  }>
}

export interface InboxItem {
  id: string
  type: 'approval' | 'proposal' | 'notification'
  title: string
  content: string
  blueprint_id?: string
  run_id?: string
  node_id?: string
  status: 'pending' | 'approved' | 'rejected' | 'read'
  created_at: string
}

export const useBlueprintStore = defineStore('blueprint', () => {
  // ── State ────────────────────────────────────
  const blueprints = ref<Blueprint[]>([])
  const currentBlueprint = ref<Blueprint | null>(null)
  const runs = ref<BlueprintRun[]>([])
  const currentRun = ref<BlueprintRun | null>(null)
  const inbox = ref<InboxItem[]>([])
  const loading = ref(false)
  const error = ref('')
  const skills = ref<SkillItem[]>([])

  // ── Skills ────────────────────────────────────
  async function fetchSkills() {
    try {
      const data = await agentJson('/v1/agent/skills')
      skills.value = Array.isArray(data) ? data : (data?.skills || data?.data || [])
    } catch (e: any) {
      // Skills fetch failure is non-critical; leave skills empty
    }
  }

  // ── Blueprint CRUD ───────────────────────────
  async function fetchBlueprints() {
    loading.value = true
    error.value = ''
    try {
      const data = await agentJson('/v1/agent/blueprints')
      blueprints.value = Array.isArray(data) ? data : (data?.blueprints || data?.data || [])
    } catch (e: any) {
      error.value = e?.message || 'Failed to fetch blueprints'
    } finally {
      loading.value = false
    }
  }

  async function createBlueprint(name: string, description: string) {
    loading.value = true
    error.value = ''
    try {
      // Backend: POST /v1/agent/blueprints — returns BlueprintDefinition directly
      const created = await agentPost('/v1/agent/blueprints', { name, description, nodes: [], edges: [] })
      blueprints.value.unshift(created)
      return created
    } catch (e: any) {
      error.value = e?.message || 'Failed to create blueprint'
      return null
    } finally {
      loading.value = false
    }
  }

  async function updateBlueprint(id: string, data: Partial<Blueprint>) {
    loading.value = true
    error.value = ''
    try {
      // Backend: PUT /v1/agent/blueprints/{id} — returns BlueprintDefinition directly
      const r = await agentFetch(`/v1/agent/blueprints/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ id, ...data }),
      })
      const updated = await r.json()
      const idx = blueprints.value.findIndex(b => b.id === id)
      if (idx >= 0) blueprints.value[idx] = { ...blueprints.value[idx], ...updated }
      if (currentBlueprint.value?.id === id) {
        currentBlueprint.value = { ...currentBlueprint.value, ...updated }
      }
      return updated
    } catch (e: any) {
      error.value = e?.message || 'Failed to update blueprint'
      return null
    } finally {
      loading.value = false
    }
  }

  async function deleteBlueprint(id: string) {
    loading.value = true
    error.value = ''
    try {
      // Backend: DELETE /v1/agent/blueprints/{id}
      await agentFetch(`/v1/agent/blueprints/${id}`, { method: 'DELETE' })
      blueprints.value = blueprints.value.filter(b => b.id !== id)
      if (currentBlueprint.value?.id === id) {
        currentBlueprint.value = null
      }
    } catch (e: any) {
      error.value = e?.message || 'Failed to delete blueprint'
    } finally {
      loading.value = false
    }
  }

  // ── Blueprint Runs ───────────────────────────
  async function fetchAllRuns() {
    loading.value = true
    error.value = ''
    try {
      const data = await agentJson('/v1/agent/runs')
      runs.value = Array.isArray(data) ? data : (data?.runs || data?.data || [])
    } catch (e: any) {
      error.value = e?.message || 'Failed to fetch runs'
    } finally {
      loading.value = false
    }
  }

  async function fetchRuns(bpId: string) {
    loading.value = true
    error.value = ''
    try {
      const data = await agentJson(`/v1/agent/blueprints/${bpId}/runs`)
      runs.value = Array.isArray(data) ? data : (data?.runs || data?.data || [])
    } catch (e: any) {
      error.value = e?.message || 'Failed to fetch runs'
    } finally {
      loading.value = false
    }
  }

  async function runBlueprint(id: string) {
    loading.value = true
    error.value = ''
    try {
      // Backend: POST /v1/agent/blueprints/{id}/runs — returns BlueprintRun directly
      const r = await agentFetch(`/v1/agent/blueprints/${id}/runs`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: '{}',
      })
      const run = await r.json()
      currentRun.value = run
      runs.value.unshift(run)
      return run
    } catch (e: any) {
      error.value = e?.message || 'Failed to run blueprint'
      return null
    } finally {
      loading.value = false
    }
  }

  async function cancelRun(id: string) {
    loading.value = true
    error.value = ''
    try {
      // Backend: POST /v1/agent/runs/{id}/cancel
      await agentFetch(`/v1/agent/runs/${id}/cancel`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: '{}',
      })
      const run = runs.value.find(r => r.id === id)
      if (run) run.status = 'cancelled'
      if (currentRun.value?.id === id) currentRun.value.status = 'cancelled'
    } catch (e: any) {
      error.value = e?.message || 'Failed to cancel run'
    } finally {
      loading.value = false
    }
  }

  // ── Inbox ────────────────────────────────────
  async function fetchInbox() {
    loading.value = true
    error.value = ''
    try {
      const data = await agentJson('/v1/agent/inbox')
      inbox.value = Array.isArray(data) ? data : (data?.items || data?.data || [])
    } catch (e: any) {
      error.value = e?.message || 'Failed to fetch inbox'
    } finally {
      loading.value = false
    }
  }

  async function approveInbox(id: string) {
    try {
      await agentPost(`/v1/agent/inbox/${id}/approve`, {})
      const item = inbox.value.find(i => i.id === id)
      if (item) item.status = 'approved'
    } catch (e: any) {
      error.value = e?.message || 'Failed to approve'
    }
  }

  async function rejectInbox(id: string) {
    try {
      await agentPost(`/v1/agent/inbox/${id}/reject`, {})
      const item = inbox.value.find(i => i.id === id)
      if (item) item.status = 'rejected'
    } catch (e: any) {
      error.value = e?.message || 'Failed to reject'
    }
  }

  async function replyInbox(id: string, text: string) {
    try {
      await agentPost(`/v1/agent/inbox/${id}/reply`, { reply: text })
      const item = inbox.value.find(i => i.id === id)
      if (item) item.status = 'approved'
    } catch (e: any) {
      error.value = e?.message || 'Failed to reply'
    }
  }

  async function markInboxRead(id: string) {
    try {
      await agentPost(`/v1/agent/inbox/${id}/read`, {})
      const item = inbox.value.find(i => i.id === id)
      if (item) item.status = 'read'
    } catch (e: any) {
      error.value = e?.message || 'Failed to mark as read'
    }
  }

  async function createInboxItem(item: { type: string; title: string; description?: string; blueprint_id?: string; run_id?: string }) {
    try {
      const result = await agentPost('/v1/agent/inbox/create', item)
      const created = result?.data || result
      inbox.value.unshift(created)
      return created
    } catch (e: any) {
      error.value = e?.message || 'Failed to create inbox item'
      return null
    }
  }

  // ── Import / Export ──────────────────────────
  async function exportBlueprint(id: string) {
    try {
      const r = await agentFetch(`/v1/agent/blueprints/${id}/export`)
      const blob = await r.blob()
      const url = URL.createObjectURL(blob)
      const a = document.createElement('a')
      a.href = url
      a.download = `blueprint-${id}.json`
      a.click()
      URL.revokeObjectURL(url)
    } catch (e: any) {
      error.value = e?.message || 'Failed to export blueprint'
    }
  }

  async function importBlueprint(file: File) {
    loading.value = true
    error.value = ''
    try {
      const text = await file.text()
      const data = JSON.parse(text)
      const result = await agentPost('/v1/agent/blueprints/import', data)
      const bp = result?.data || result
      blueprints.value.unshift(bp)
      return bp
    } catch (e: any) {
      error.value = e?.message || 'Failed to import blueprint'
      return null
    } finally {
      loading.value = false
    }
  }

  // ── Select / Deselect ────────────────────────
  function selectBlueprint(bp: Blueprint | null) {
    currentBlueprint.value = bp
  }

  function selectRun(run: BlueprintRun | null) {
    currentRun.value = run
  }

  return {
    // State
    blueprints, currentBlueprint, runs, currentRun, inbox,
    loading, error, skills,
    // Blueprint CRUD
    fetchBlueprints, createBlueprint, updateBlueprint, deleteBlueprint,
    // Runs
    fetchAllRuns, fetchRuns, runBlueprint, cancelRun,
    // Inbox
    fetchInbox, approveInbox, rejectInbox, replyInbox, markInboxRead, createInboxItem,
    // Import / Export
    exportBlueprint, importBlueprint,
    // Selection
    selectBlueprint, selectRun,
    // Skills
    fetchSkills,
  }
})

File: blueprint.go
package main

import "time"

// ── Blueprint Definition ────────────────────────────────────────────────

// BlueprintDefinition represents a saved blueprint workflow (a DAG of nodes and edges).
type BlueprintDefinition struct {
	ID          string                 `json:"id"`
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	Nodes       []BlueprintNode        `json:"nodes"`
	Edges       []BlueprintEdge        `json:"edges"`
	Variables   map[string]interface{} `json:"variables"`
	Display     BlueprintDisplay       `json:"display"`
	CreatedAt   time.Time              `json:"created_at"`
	UpdatedAt   time.Time              `json:"updated_at"`
	Version     int                    `json:"version"`
}

// BlueprintDisplay holds viewport and other display-related settings.
type BlueprintDisplay struct {
	Viewport BlueprintViewport `json:"viewport"`
}

// BlueprintViewport represents the camera/zoom state for the visual editor.
type BlueprintViewport struct {
	X    float64 `json:"x"`
	Y    float64 `json:"y"`
	Zoom float64 `json:"zoom"`
}

// ── Node ────────────────────────────────────────────────────────────────

// NodeType enumerates the kinds of nodes that can appear in a blueprint.
type NodeType string

const (
	NodeTypeAgent       NodeType = "agent"
	NodeTypeManager     NodeType = "manager"
	NodeTypeManagerSlot NodeType = "manager_slot"
	NodeTypeGroup       NodeType = "group"
	NodeTypeCondition   NodeType = "condition"
	NodeTypeLoop        NodeType = "loop"
	NodeTypeSummary     NodeType = "summary"
	NodeTypeNote        NodeType = "note"
)

// ResultRole determines how a node's output is treated in the final result.
type ResultRole string

const (
	ResultRoleAuto   ResultRole = "auto"
	ResultRoleFinal  ResultRole = "final"
	ResultRoleIgnore ResultRole = "ignore"
)

// BlueprintNode is a single step in a blueprint DAG.
type BlueprintNode struct {
	ID         string                 `json:"id"`
	Type       NodeType               `json:"type"`
	Label      string                 `json:"label"`
	ParentID   string                 `json:"parent_id"`
	Disabled   bool                   `json:"disabled"`
	Position   BlueprintNodePosition  `json:"position"`
	Size       BlueprintNodeSize      `json:"size"`
	RuntimeID  string                 `json:"runtime_id"`
	ResultRole ResultRole             `json:"result_role"`
	Config     map[string]interface{} `json:"config"`
}

// BlueprintNodePosition stores the x/y coordinates for the visual editor.
type BlueprintNodePosition struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// BlueprintNodeSize stores the width/height dimensions for the visual editor.
type BlueprintNodeSize struct {
	Width  float64 `json:"width"`
	Height float64 `json:"height"`
}

// ── Node Config Structs ────────────────────────────────────────────────

// AgentApprovalConfig configures the approval gate on an agent node.
type AgentApprovalConfig struct {
	Enabled bool `json:"enabled"`
}

// AgentSendConfig configures the send/action behaviour on an agent node.
type AgentSendConfig struct {
	Enabled      bool   `json:"enabled"`
	ChannelID    string `json:"channel_id"`
	Target       string `json:"target"`
	BodyTemplate string `json:"body_template"`
}

// AgentConfig holds the typed configuration for an agent node.
type AgentConfig struct {
	RuntimeID         string              `json:"runtime_id"`
	AgentName         string              `json:"agent_name"`
	Prompt            string              `json:"prompt"`
	SkillIDs          []string            `json:"skill_ids"`
	ModelID           string              `json:"model_id"`
	PermissionProfile string              `json:"permission_profile"`
	WorkingDirectory  string              `json:"working_directory"`
	TimeoutMs         int                 `json:"timeout_ms"`
	OutputSchema      interface{}         `json:"output_schema"`
	Approval          AgentApprovalConfig `json:"approval"`
	Send              AgentSendConfig     `json:"send"`
	Tools             []string            `json:"tools"`
}

// ManagerConfig holds the typed configuration for a manager node.
type ManagerConfig struct {
	PortCount         int      `json:"port_count"`
	MaxHandoffs       int      `json:"max_handoffs"`
	Instructions      string   `json:"instructions"`
	RuntimeID         string   `json:"runtime_id"`
	AgentName         string   `json:"agent_name"`
	ModelID           string   `json:"model_id"`
	SkillIDs          []string `json:"skill_ids"`
	PermissionProfile string   `json:"permission_profile"`
	WorkingDirectory  string   `json:"working_directory"`
	TimeoutMs         int      `json:"timeout_ms"`
	Tools             []string `json:"tools"`
}

// ExecutionMode determines how a manager slot executes its children.
type ExecutionMode string

const (
	ExecutionModeManual   ExecutionMode = "manual"
	ExecutionModeParallel ExecutionMode = "parallel"
)

// ManagerSlotConfig holds the typed configuration for a manager_slot node.
type ManagerSlotConfig struct {
	ManagerNodeID     string        `json:"manager_node_id"`
	Slot              int           `json:"slot"`
	ExecutionMode     ExecutionMode `json:"execution_mode"`
	ParallelLaneCount int           `json:"parallel_lane_count"`
}

// SummaryMode determines how a summary node merges its inputs.
type SummaryMode string

const (
	SummaryModeStructuredMerge SummaryMode = "structured_merge"
	SummaryModeHarnessSummary  SummaryMode = "harness_summary"
)

// SummaryConfig holds the typed configuration for a summary node.
type SummaryConfig struct {
	Mode      SummaryMode `json:"mode"`
	RuntimeID string      `json:"runtime_id"`
	Prompt    string      `json:"prompt"`
	ModelID   string      `json:"model_id"`
}

// GroupConfig holds the typed configuration for a group node.
type GroupConfig struct {
	Color string `json:"color"`
}

// ConditionConfig holds the typed configuration for a condition node.
type ConditionConfig struct {
	Expression string `json:"expression"`
}

// LoopConfig holds the typed configuration for a loop node.
type LoopConfig struct {
	MaxIterations int `json:"max_iterations"`
}

// ── Edge ────────────────────────────────────────────────────────────────

// BlueprintEdge connects two nodes in a blueprint DAG.
type BlueprintEdge struct {
	ID           string `json:"id"`
	Source       string `json:"source"`
	Target       string `json:"target"`
	SourceHandle string `json:"source_handle"`
	TargetHandle string `json:"target_handle"`
	Condition    string `json:"condition"`
	Label        string `json:"label"`
}

// ── Run ─────────────────────────────────────────────────────────────────

// BlueprintRunStatus enumerates the possible states of a blueprint execution.
type BlueprintRunStatus string

const (
	RunStatusQueued          BlueprintRunStatus = "queued"
	RunStatusRunning         BlueprintRunStatus = "running"
	RunStatusSucceeded       BlueprintRunStatus = "succeeded"
	RunStatusFailed          BlueprintRunStatus = "failed"
	RunStatusCancelled       BlueprintRunStatus = "cancelled"
	RunStatusWaitingApproval BlueprintRunStatus = "waiting_approval"
)

// BlueprintRun tracks a single execution of a blueprint.
type BlueprintRun struct {
	ID                string                `json:"id"`
	BlueprintID       string                `json:"blueprint_id"`
	BlueprintName     string                `json:"blueprint_name"`
	BlueprintVersion  int                   `json:"blueprint_version"`
	StartedBy         string                `json:"started_by"`
	Status            BlueprintRunStatus    `json:"status"`
	NodeRuns          []BlueprintNodeRun    `json:"node_runs"`
	Events            []BlueprintNodeEvent  `json:"events"`
	StartedAt         *time.Time            `json:"started_at"`
	FinishedAt        *time.Time            `json:"finished_at"`
	DurationMs        int64                 `json:"duration_ms"`
	FinalResult       interface{}           `json:"final_result"`
	TotalInputTokens  int64                 `json:"total_input_tokens"`
	TotalOutputTokens int64                 `json:"total_output_tokens"`
	TotalCostUsd      float64               `json:"total_cost_usd"`
	TokenUsage        int                   `json:"token_usage"`
}

// ── Node Run ────────────────────────────────────────────────────────────

// BlueprintNodeRunStatus enumerates the possible states of a single node execution.
type BlueprintNodeRunStatus string

const (
	NodeRunStatusPending         BlueprintNodeRunStatus = "pending"
	NodeRunStatusQueued          BlueprintNodeRunStatus = "queued"
	NodeRunStatusRunning         BlueprintNodeRunStatus = "running"
	NodeRunStatusSucceeded       BlueprintNodeRunStatus = "succeeded"
	NodeRunStatusFailed          BlueprintNodeRunStatus = "failed"
	NodeRunStatusSkipped         BlueprintNodeRunStatus = "skipped"
	NodeRunStatusCancelled       BlueprintNodeRunStatus = "cancelled"
	NodeRunStatusWaitingApproval BlueprintNodeRunStatus = "waiting_approval"
)

// NodeRunUsage tracks token consumption for a single node execution.
type NodeRunUsage struct {
	PromptTokens     int64 `json:"prompt_tokens"`
	CompletionTokens int64 `json:"completion_tokens"`
}

// BlueprintNodeRun tracks the execution of a single node within a blueprint run.
type BlueprintNodeRun struct {
	ID             string                  `json:"id"`
	BlueprintRunID string                 `json:"blueprint_run_id"`
	BlueprintID    string                  `json:"blueprint_id"`
	NodeID         string                  `json:"node_id"`
	NodeLabel      string                  `json:"node_label"`
	NodeType       NodeType                `json:"node_type"`
	Status         BlueprintNodeRunStatus  `json:"status"`
	QueuedAt       *time.Time              `json:"queued_at"`
	Input          interface{}             `json:"input"`
	Output         interface{}             `json:"output"`
	Error          string                  `json:"error"`
	StartedAt      *time.Time              `json:"started_at"`
	FinishedAt     *time.Time              `json:"finished_at"`
	Usage          NodeRunUsage            `json:"usage"`
	OpenClawRef    string                  `json:"openclaw_ref"`
	TokenUsage     int                     `json:"token_usage"`
	ApprovalState  *ApprovalState          `json:"approval_state"`
}

// ── Events ──────────────────────────────────────────────────────────────

// BlueprintEventKind enumerates the event types emitted during a blueprint run.
type BlueprintEventKind string

const (
	EventBlueprintRunStarted    BlueprintEventKind = "blueprint.run.started"
	EventNodeRunQueued          BlueprintEventKind = "node.run.queued"
	EventNodeRunStarted         BlueprintEventKind = "node.run.started"
	EventNodeRunWaitingApproval BlueprintEventKind = "node.run.waiting_approval"
	EventNodeRunCompleted       BlueprintEventKind = "node.run.completed"
	EventNodeRunFailed          BlueprintEventKind = "node.run.failed"
	EventNodeRunCancelled       BlueprintEventKind = "node.run.cancelled"
	EventBlueprintRunCompleted  BlueprintEventKind = "blueprint.run.completed"
	EventBlueprintRunCancelled  BlueprintEventKind = "blueprint.run.cancelled"
	EventBlueprintRunFailed     BlueprintEventKind = "blueprint.run.failed"
)

// BlueprintNodeEvent represents a single event emitted during a blueprint run.
type BlueprintNodeEvent struct {
	ID             string             `json:"id"`
	BlueprintRunID string             `json:"blueprint_run_id"`
	NodeRunID      string             `json:"node_run_id,omitempty"`
	Type           BlueprintEventKind `json:"type"`
	Message        string             `json:"message"`
	CreatedAt      time.Time          `json:"created_at"`
}

// ── Final Result ────────────────────────────────────────────────────────

// FinalRunResultState enumerates the possible states of a blueprint run's final result.
type FinalRunResultState string

const (
	FinalRunResultStateAvailable       FinalRunResultState = "available"
	FinalRunResultStateFailed          FinalRunResultState = "failed"
	FinalRunResultStateWaitingApproval FinalRunResultState = "waiting_approval"
	FinalRunResultStateEmpty           FinalRunResultState = "empty"
)

// SelectionReason enumerates the reasons a candidate was selected as the final result.
type SelectionReason string

const (
	SelectionReasonExplicitFinal  SelectionReason = "explicit_final"
	SelectionReasonTerminalResult SelectionReason = "terminal_result"
	SelectionReasonLatestResult   SelectionReason = "latest_result"
)

// FinalRunResult captures the outcome of a completed blueprint run.
type FinalRunResult struct {
	State               FinalRunResultState       `json:"state"`
	Candidates          []FinalRunResultCandidate `json:"candidates"`
	FailedNode          *FinalRunResultCandidate  `json:"failed_node,omitempty"`
	WaitingApprovalNode *FinalRunResultCandidate  `json:"waiting_approval_node,omitempty"`
}

// FinalRunResultCandidate represents a single candidate output in the final result.
type FinalRunResultCandidate struct {
	NodeRunID       string          `json:"node_run_id"`
	BlueprintRunID  string          `json:"blueprint_run_id"`
	BlueprintID     string          `json:"blueprint_id"`
	NodeID          string          `json:"node_id"`
	NodeLabel       string          `json:"node_label"`
	NodeType        NodeType        `json:"node_type"`
	ResultRole      ResultRole      `json:"result_role"`
	SelectionReason SelectionReason `json:"selection_reason"`
	Output          interface{}     `json:"output"`
	EndedAt         *time.Time      `json:"ended_at"`
}

// ── Approval ────────────────────────────────────────────────────────────

// ApprovalStatus enumerates the possible states of an approval request.
type ApprovalStatus string

const (
	ApprovalStatusPending  ApprovalStatus = "pending"
	ApprovalStatusApproved ApprovalStatus = "approved"
	ApprovalStatusDenied   ApprovalStatus = "denied"
	ApprovalStatusTimeout  ApprovalStatus = "timeout"
)

// ApprovalType enumerates the kinds of approval gates.
type ApprovalType string

const (
	ApprovalTypeHuman ApprovalType = "human"
)

// ApprovalState captures the state of a human-approval gate.
type ApprovalState struct {
	RequestID       string         `json:"request_id"`
	Status          ApprovalStatus `json:"status"`
	Replies         []string       `json:"replies"`
	SelectedReplyID string         `json:"selected_reply_id"`
	SelectedOutput  interface{}    `json:"selected_output"`
	ReviewOutput    interface{}    `json:"review_output"`
	ApprovalType    ApprovalType   `json:"approval_type"`
}

// ── Inbox ───────────────────────────────────────────────────────────────

// InboxItemType enumerates the kinds of items that can appear in the user inbox.
type InboxItemType string

const (
	InboxItemTypeApproval     InboxItemType = "approval"
	InboxItemTypeProposal     InboxItemType = "proposal"
	InboxItemTypeNotification InboxItemType = "notification"
)

// InboxItemStatus enumerates the possible states of an inbox item.
type InboxItemStatus string

const (
	InboxItemStatusPending   InboxItemStatus = "pending"
	InboxItemStatusApproved  InboxItemStatus = "approved"
	InboxItemStatusRejected  InboxItemStatus = "rejected"
	InboxItemStatusRead      InboxItemStatus = "read"
	InboxItemStatusDone      InboxItemStatus = "done"
)

// InboxItem represents an item in the user's approval/notification inbox.
type InboxItem struct {
	ID               string                    `json:"id"`
	Type             InboxItemType             `json:"type"`
	Title            string                    `json:"title"`
	Description      string                    `json:"description"`
	Status           InboxItemStatus           `json:"status"`
	BlueprintID      string                    `json:"blueprint_id"`
	RunID            string                    `json:"run_id"`
	NodeID           string                    `json:"node_id"`
	Comment          string                    `json:"comment"`
	BlueprintPackage *PortableBlueprintPackage `json:"blueprint_package,omitempty"`
	CreatedByRoleID  string                    `json:"created_by_role_id"`
	TargetRoleID     string                    `json:"target_role_id"`
	Summary          string                    `json:"summary"`
	Preview          string                    `json:"preview"`
	DiffSummary      string                    `json:"diff_summary"`
	Schema           interface{}               `json:"schema"`
	CreatedAt        time.Time                 `json:"created_at"`
	ResolvedAt       *time.Time                `json:"resolved_at"`
}

// ── Import / Export ─────────────────────────────────────────────────────

// PortableBlueprintPackage is the serialisation format for blueprint import/export.
// Schema is always "hixns.blueprint-package/v1". On export, companyId and timestamps
// are stripped, send.channelId/target/tools are cleared. On import, auto-layout is applied.
type PortableBlueprintPackage struct {
	Schema     string                `json:"schema"`
	ExportedAt time.Time             `json:"exported_at"`
	Blueprints []BlueprintDefinition `json:"blueprints"`
}
